### ISSUE TITLE
[Python] Async API ServicerContext cancelled() returning False during callback

### ISSUE THREAD

**Issue by purujitgoyal** [grpc/grpc#36193]
### What version of gRPC and what language are you using?
grpcio: 1.62.1
grpcio-tools: 1.62.1

### What operating system (Linux, Windows,...) and version?
Linux

### What runtime / compiler are you using (e.g. python version or version of gcc)
Python 3.11

### What did you do?
I am working with the async api and having problems with handling client cancellations on the server side. Here's a small prototype built using [this example](https://github.com/grpc/grpc/tree/master/examples/python/hellostreamingworld) describing the problem:
#### async_server.py
```
def callback(context: grpc.aio.ServicerContext):
    logging.info("Inside callback")
    logging.info(context.cancelled())


class Greeter(MultiGreeterServicer):
    async def sayHello(
        self, request: HelloRequest, context: grpc.aio.ServicerContext
    ) -> HelloReply:
        logging.info("Serving sayHello request %s", request)
        context.add_done_callback(callback)
        try:
            for i in range(10):
                await asyncio.sleep(1)
                yield HelloReply(message=f"Hello number {i}, {request.name}!")
        except asyncio.CancelledError:
            logging.info("RPC cancelled")
        finally:
            await asyncio.sleep(2)
            logging.info("Inside finally")
```

#### async_client.py
```
async def run() -> None:
    async with grpc.aio.insecure_channel("localhost:50051") as channel:
        stub = hellostreamingworld_pb2_grpc.MultiGreeterStub(channel)
        hello_stream = stub.sayHello(hellostreamingworld_pb2.HelloRequest(name="you"))
        try:
            while True:
                response = await hello_stream.read()
                if response == grpc.aio.EOF:
                    break
                logging.info("Greeter client received from direct read: " + response.message)
                await asyncio.sleep(2)
                hello_stream.cancel()
        except asyncio.CancelledError:
            pass


if __name__ == "__main__":
    logging.basicConfig()
    asyncio.run(run())
```
### What did you expect to see?
`context.cancelled()` should return True inside the callback function

### What did you see instead?
`context.cancelled()` returns False inside the callback function

### Anything else we should know about your project / environment?

Separate question, why does the await statement inside the finally block not throw the CancelledError like the one in the try block? What is the preferred way to handle cancellations from client side, catching the Cancelled Error and doing cleanup in `finally` or through callback function?

**Comment by purujitgoyal** [comment]
Yes, it's the same issue and I used the approach mentioned in the SO answer, but that answer was before the `cancelled` and `done` methods were introduced in #27767.

Anyway, what would you recommend for the cleanup in this case? The cleanup is an async function as well.

1. `await` it inside the finally block
2. Use `asyncio.create_task` in `add_done_callback`

**Comment by XuanWang-Amos** [comment]
Both looks fine to me so I guess the answer is depends on your user case.

**Comment by stansotn** [comment]
I am experiencing the same issue with the python async API streaming RPCs. Following are my observations:

1. Regarding `context.add_done_callback(callback)` in the server handler. The callback is reliably called both when the streaming rpc is cancelled by client and when the streaming rpc completes. The server's `context.cancelled()` always returns `False` no matter whether the rpc had been cancelled or completed. The interesting part is that `context.done()` returns `True` only when there is a delay in the server handler (`await asyncio.sleep(1)` in the example above). If we remove the sleep call, the callback still gets called when client cancels, but `context.done()` is set to `False`.
2. In the example provided by the author, the `asyncio.CancelledError` cancelled error is only observed with the delay (`await asyncio.sleep(1)`). Again, the callback seems to get called properly without the delay with `.done()` and `.cancelled()` flags set to `False`, but the exception is not being raised.

@purujitgoyal Were you able to come up with a workaround to reliably detect if streaming call is being cancelled by client?

**Comment by stansotn** [comment]
I suppose the `context.done()` flag in the "done_callback" can be relied upon for determining whether the rpc had been cancelled by the client. After some tinkering I can affirm that the `context.done()` flag is reliably set when rpc handler runs to completion. The only reason it may not run to completion is being cancelled by the client or by the server (in which case the server would be aware by other means). 

The fact that the `context.cancelled()` flag is `False` in the callback when rpc is cancelled is quite confusing (bug?) given the lack of documentation on this matter.

**Comment by AltairLyx** [comment]
> Hi, looks like it's the same issue mentioned here: https://stackoverflow.com/questions/68491834/handle-client-side-cancellation-in-grpc-python-asyncio
> 
> In short, the client-side-cancellation is observed as an `asyncio.CancelledError` in method handler. Currently we recommend use `except asyncio.CancelledError` to caught and handle client cancel on server side.

is it by design?
i think it's more reasonable that the service_context's cancelled method returns true if request is cancelled by client in callback functions when request is terminated.

**Comment by AltairLyx** [comment]
@gnossen @XuanWang-Amos

**Comment by metasyn** [comment]
👋 Hello - I just ran into this issue today. Is it clear what to do to fix this? I am happy to try to submit a fix, but I don't feel its obvious from the maintainers as to what the fix might be. @XuanWang-Amos, or @gnossen,  could you advise here perhaps? Thanks!
